README
This code is a clone of the classic computer game Minesweeper. The game is a 24x12 grid, some of which are hiding bombs.

Non-bomb tiles will show the number of bombs in the 8 tiles surrounding it. Left click a tile to reveal it. Right click a tile to flag it as a bomb. If a tile has the correct number of flags surrounding it, you can middle click it reveal its remaining neighboring tiles.

Additionally, when a tile is middle clicked and held, all surrounding unclicked tiles are highlighted.

The game is won when all bombs are flagged, and is lost when a bomb is left clicked.

There are three modes, the difficulty of which is determined by how many bombs are in the grid. Easy mode has 20 bombs, medium has 30
bombs, and hard has 50 bombs.

by alix noble, spring 2023 for com s 327 at iowa state

CHANGELOG
04/19/23    04:57PM     Can draw a 20x12 grid that fills the terminal, each tile has room for a single character.

            05:29PM     Can now put things in the grid without redrawing the grid!!

04/20/23    02:52PM     Boards can be generated with varying numbers of bombs and be printed into the grid.

            03:51PM     Board generation complete! non-bomb tiles know how many bombs they border, grid is printed nice and in color

04/24/23    03:20PM     Implemented mouse input for left, right, and middle click. When a tile is left clicked, its value is revealed.
                        Currently the game does not end when a bomb is hit and right and middle click don't do anything useful.

04/25/23    11:22AM     When a zero is revealed, all of the tiles around it are also revealed, game ends when bomb is clicked

            03:54PM     middle clicking, flagging, win conditions

04/30/23    03:32PM     title screen

05/01/23    02:37PM     first click will now never be a bomb

            08:16PM     difficulty selection, option to play another game after one is finished

05/05/23    02:23PM     when a tile is middle clicked and held, all surrounding unclicked tiles are highlighted

tile.h
#ifndef TILE_H
#define TILE_H

#include <vector>

#define BOARD_X 20
#define BOARD_Y 12

#define in_bounds(x, y) (x > -1 && x < BOARD_X && y > -1 && y < BOARD_Y)

class tile {
private:
    bool is_bomb;
    int value;
    int x;
    int y;

public:
    bool is_clicked;
    bool is_flagged;
    bool is_highlighted;
    tile(bool is_bomb, int value, int x, int y) : is_bomb(is_bomb), value(value), x(x), y(y), is_clicked(false), is_flagged(false), is_highlighted(false) {}
    bool check_if_bomb() { return is_bomb; }
    int get_value() { return value; }
    int get_x() { return x; }
    int get_y() { return y; }
    ~tile() {}
};

class game_board {
private:
    int num_bombs;
    std::vector<tile *> bombs;
    int num_flags;
public:
    tile *board[BOARD_Y][BOARD_X];

    void increment_flags() { num_flags++; }
    void decrement_flags() { if (num_flags > 0) { num_flags--; } }
    int get_num_flags() { return num_flags; }

    std::vector<tile *> get_tile_neighbors(int x, int y);
    int check_tile_neighbors_bombs(int x, int y);
    int check_tile_neighbors_flags(int x, int y);

    bool have_won();
    
    game_board(int num_bombs, int click_x, int click_y);
    ~game_board();
};

#endif

tile.cpp
#include "tile.h"
#include <cstdlib>
#include <ctime>
#include <cstdio>
#include <array>
#include <algorithm>
#include <ncurses.h>

int game_board::check_tile_neighbors_bombs(int x, int y) {
    int num_bombs = 0;
    if (in_bounds(x - 1, y - 1) && board[y - 1][x - 1] && board[y - 1][x - 1]->check_if_bomb()) {
        num_bombs++;
    }
    if (in_bounds(x, y - 1)     && board[y - 1][x    ] && board[y - 1][x    ]->check_if_bomb()) {
        num_bombs++;
    }
    if (in_bounds(x + 1, y - 1) && board[y - 1][x + 1] && board[y - 1][x + 1]->check_if_bomb()) {
        num_bombs++;
    }
    if (in_bounds(x - 1, y)     && board[y    ][x - 1] && board[y    ][x - 1]->check_if_bomb()) {
        num_bombs++;
    }
    if (in_bounds(x + 1, y)     && board[y    ][x + 1] && board[y    ][x + 1]->check_if_bomb()) {
        num_bombs++;
    }
    if (in_bounds(x - 1, y + 1) && board[y + 1][x - 1] && board[y + 1][x - 1]->check_if_bomb()) {
        num_bombs++;
    }
    if (in_bounds(x, y + 1)     && board[y + 1][x    ] && board[y + 1][x    ]->check_if_bomb()) {
        num_bombs++;
    }
    if (in_bounds(x + 1, y + 1) && board[y + 1][x + 1] && board[y + 1][x + 1]->check_if_bomb()) {
        num_bombs++;
    }
    return num_bombs;
}

int game_board::check_tile_neighbors_flags(int x, int y) {
    int num_flags = 0;
    if (in_bounds(x - 1, y - 1) && board[y - 1][x - 1] && board[y - 1][x - 1]->is_flagged) {
        num_flags++;
    }
    if (in_bounds(x, y - 1)     && board[y - 1][x    ] && board[y - 1][x    ]->is_flagged) {
        num_flags++;
    }
    if (in_bounds(x + 1, y - 1) && board[y - 1][x + 1] && board[y - 1][x + 1]->is_flagged) {
        num_flags++;
    }
    if (in_bounds(x - 1, y)     && board[y    ][x - 1] && board[y    ][x - 1]->is_flagged) {
        num_flags++;
    }
    if (in_bounds(x + 1, y)     && board[y    ][x + 1] && board[y    ][x + 1]->is_flagged) {
        num_flags++;
    }
    if (in_bounds(x - 1, y + 1) && board[y + 1][x - 1] && board[y + 1][x - 1]->is_flagged) {
        num_flags++;
    }
    if (in_bounds(x, y + 1)     && board[y + 1][x    ] && board[y + 1][x    ]->is_flagged) {
        num_flags++;
    }
    if (in_bounds(x + 1, y + 1) && board[y + 1][x + 1] && board[y + 1][x + 1]->is_flagged) {
        num_flags++;
    }
    return num_flags;
}

std::vector<tile *> game_board::get_tile_neighbors(int x, int y) {
    std::vector<tile *> v;
    if (in_bounds(x - 1, y - 1) && board[y - 1][x - 1]) {
        v.push_back(board[y - 1][x - 1]);
    }
    if (in_bounds(x, y - 1)     && board[y - 1][x    ]) {
        v.push_back(board[y - 1][x    ]);
    }
    if (in_bounds(x + 1, y - 1) && board[y - 1][x + 1]) {
        v.push_back(board[y - 1][x + 1]);
    }
    if (in_bounds(x - 1, y)     && board[y    ][x - 1]) {
        v.push_back(board[y    ][x - 1]);
    }
    if (in_bounds(x + 1, y)     && board[y    ][x + 1]) {
        v.push_back(board[y    ][x + 1]);
    }
    if (in_bounds(x - 1, y + 1) && board[y + 1][x - 1]) {
        v.push_back(board[y + 1][x - 1]);
    }
    if (in_bounds(x, y + 1)     && board[y + 1][x    ]) {
        v.push_back(board[y + 1][x    ]);
    }
    if (in_bounds(x + 1, y + 1) && board[y + 1][x + 1]) {
        v.push_back(board[y + 1][x + 1]);
    }
    return v;
}

bool game_board::have_won() {
    if (num_bombs != num_flags) {
        return false;
    }

    for (tile *t : bombs) {
        if (!t->is_flagged) {
            return false;
        }
    }
    return true;
}

game_board::game_board(int num_bombs, int click_x, int click_y) : num_bombs(num_bombs), num_flags(0) {
    int x, y;
    int n = 0;
    std::vector<std::array<int, 2>> reserved;

    if (in_bounds(click_x, click_y)) {
        reserved.push_back({click_y, click_x});
    }
    if (in_bounds(click_x - 1, click_y - 1)) {
        reserved.push_back({click_y - 1, click_x - 1});
    }
    if (in_bounds(click_x, click_y - 1)) {
        reserved.push_back({click_y - 1, click_x    });
    }
    if (in_bounds(click_x + 1, click_y - 1)) {
        reserved.push_back({click_y - 1, click_x + 1});
    }
    if (in_bounds(click_x - 1, click_y)) {
        reserved.push_back({click_y    , click_x - 1});
    }
    if (in_bounds(click_x + 1, click_y)) {
        reserved.push_back({click_y    , click_x + 1});
    }
    if (in_bounds(click_x - 1, click_y + 1)) {
        reserved.push_back({click_y + 1, click_x - 1});
    }
    if (in_bounds(click_x, click_y + 1)) {
        reserved.push_back({click_y + 1, click_x    });
    }
    if (in_bounds(click_x + 1, click_y + 1)) {
        reserved.push_back({click_y + 1, click_x + 1});
    }

    for (y = 0; y < BOARD_Y; ++y) {
        for (x = 0; x < BOARD_X; ++x) {
            board[y][x] = NULL;
        }
    }
    

    srand(time(NULL));

    while (n < num_bombs) {
        x = rand() % BOARD_X;
        y = rand() % BOARD_Y;
        if (!board[y][x] && std::find(reserved.begin(), reserved.end(), std::array<int, 2> {y, x}) == reserved.end()) {
            board[y][x] = new tile(true, -1, x, y);
            bombs.push_back(board[y][x]);
            //mvprintw(26, 0, "%d: (%d, %d)\n", n, x, y);
            n++;
        }
    }

    for (y = 0; y < BOARD_Y; ++y) {
        for (x = 0; x < BOARD_X; ++x) {
            if (!board[y][x]) {
                board[y][x] = new tile(false, check_tile_neighbors_bombs(x, y), x, y);
            }
        }
    }
}

game_board::~game_board() {
    int x, y;

    for (y = 0; y < BOARD_Y; ++y) {
        for (x = 0; x < BOARD_X; ++x) {
            delete board[y][x];
        }
    }
}

minesweeper.cpp
#include <iostream>
#include <ncurses.h>
#include <vector>
#include <cmath>

#include "tile.h"

#define EASY    20
#define MEDIUM  30
#define HARD    50

#define io_x(x) (x * 4 + 2)
#define io_y(y) (y * 2 + 1)
#define x(io_x) ((int)round(((double)(io_x - 2) / 4.0)))
#define y(io_y) ((int)round(((double)(io_y - 1) / 2.0)))

int colors[9] = 
{
  COLOR_BLACK,  //bomb
  COLOR_BLUE,   //1
  COLOR_GREEN,  //2
  COLOR_RED,    //3
  COLOR_MAGENTA,//4
  COLOR_CYAN,   //5
  COLOR_YELLOW, //6
  COLOR_WHITE,  //7
  COLOR_RED     //8
};

void init_terminal(void)
{
  initscr();
  raw();
  noecho();
  curs_set(0);
  keypad(stdscr, TRUE);
  mousemask(ALL_MOUSE_EVENTS | REPORT_MOUSE_POSITION, NULL);
  //mousemask(REPORT_MOUSE_POSITION, NULL);
  start_color();
  init_pair(COLOR_RED, COLOR_RED, COLOR_BLACK);
  init_pair(COLOR_GREEN, COLOR_GREEN, COLOR_BLACK);
  init_pair(COLOR_YELLOW, COLOR_YELLOW, COLOR_BLACK);
  init_pair(COLOR_BLUE, COLOR_BLUE, COLOR_BLACK);
  init_pair(COLOR_MAGENTA, COLOR_MAGENTA, COLOR_BLACK);
  init_pair(COLOR_CYAN, COLOR_CYAN, COLOR_BLACK);
  init_pair(COLOR_WHITE, COLOR_WHITE, COLOR_BLACK);
}

void reset_terminal(void)
{
  endwin();
}

void draw_grid(void) {
    //attron(COLOR_PAIR(COLOR_YELLOW));
    //left side
    mvaddch(1, 0, ACS_VLINE);
    mvaddch(0, 0, ACS_ULCORNER);
    mvaddch(0, 1, ACS_HLINE);
    mvaddch(0, 2, ACS_HLINE);
    mvaddch(0, 3, ACS_HLINE);
    //mvaddch(1, 2, 'a');

    for (int y = 2; y < 24; y += 2) {
      mvaddch(y + 1, 0, ACS_VLINE);
      mvaddch(y    , 0, ACS_LTEE);
      mvaddch(y    , 1, ACS_HLINE);
      mvaddch(y    , 2, ACS_HLINE);
      mvaddch(y    , 3, ACS_HLINE);
      //mvaddch(y + 1, 2, 'a');
    }

    mvaddch(24, 0, ACS_LLCORNER);
    mvaddch(24, 1, ACS_HLINE);
    mvaddch(24, 2, ACS_HLINE);
    mvaddch(24, 3, ACS_HLINE);

    //right side
    mvaddch(0, 80, ACS_URCORNER);
    mvaddch(1, 80, ACS_VLINE);
    for (int y = 2; y < 24; y += 2) {
      mvaddch(y + 1, 80, ACS_VLINE);
      mvaddch(y    , 80, ACS_RTEE);
    }
    mvaddch(24, 80, ACS_LRCORNER);
    //attroff(COLOR_PAIR(COLOR_YELLOW));

    
    //attron(COLOR_PAIR(COLOR_GREEN));
    //top
    for (int x = 4; x < 80; x += 4) {
      mvaddch(1, x    , ACS_VLINE);
      mvaddch(0, x    , ACS_TTEE);
      mvaddch(0, x + 1, ACS_HLINE);
      mvaddch(0, x + 2, ACS_HLINE);
      mvaddch(0, x + 3, ACS_HLINE);
      //mvaddch(1, x + 2, 'a');
    }

    //bottom
    for (int x = 4; x < 80; x += 4) {
      mvaddch(24, x    , ACS_BTEE);
      mvaddch(24, x + 1, ACS_HLINE);
      mvaddch(24, x + 2, ACS_HLINE);
      mvaddch(24, x + 3, ACS_HLINE);
    }
    //attroff(COLOR_PAIR(COLOR_GREEN));


    //inner
    for (int y = 2; y < 24; y += 2) {
      for (int x = 4; x < 80; x += 4) {
        mvaddch(y + 1, x    , ACS_VLINE);
        mvaddch(y    , x    , ACS_PLUS);
        mvaddch(y    , x + 1, ACS_HLINE);
        mvaddch(y    , x + 2, ACS_HLINE);
        mvaddch(y    , x + 3, ACS_HLINE);
        //mvaddch(y + 1, x + 2, 'a');
      }
    }
  
}

void fill_tile(game_board *board, int x, int y) {
  int color;
  if (board->board[y][x]->get_value() != 0) {
    if (board->board[y][x]->check_if_bomb()) {
      color = colors[0];
    } else {
      color = colors[board->board[y][x]->get_value()];
    }
    attron(COLOR_PAIR(color));
    mvaddch(io_y(y), io_x(x), board->board[y][x]->check_if_bomb() ? '<' : board->board[y][x]->get_value() + '0');
    attroff(COLOR_PAIR(color));
    refresh();
  }
  else {
    mvaddch(io_y(y), io_x(x), ' ');
  }
}

void show_grid(game_board *board) {
  for (int y = 0; y < BOARD_Y; ++y) {
    for (int x = 0; x < BOARD_X; ++x) {
      fill_tile(board, x, y);
    }
  }
}

void grid_init() {
  clear();
  draw_grid();
  for (int y = 0; y < BOARD_Y; ++y) {
    for (int x = 0; x < BOARD_X; ++x) {
      attron(COLOR_PAIR(COLOR_CYAN));
      mvaddch(io_y(y), io_x(x), '.');
      attroff(COLOR_PAIR(COLOR_CYAN));
      refresh();
    }
  }
}

bool on_tile_click(game_board *board, int x, int y) {
  fill_tile(board, x, y);
  board->board[y][x]->is_clicked = true;
  board->board[y][x]->is_flagged = false;
  if (board->board[y][x]->get_value() == 0) {
    std::vector<tile *> v = board->get_tile_neighbors(x, y);
    for (tile *t : v) {
      if (!t->is_clicked) {
        on_tile_click(board, t->get_x(), t->get_y());
      }
    }
  }
  if (board->board[y][x]->check_if_bomb()) {
    return true;
  }
  return false;
}

void on_tile_flag(game_board *b, tile *t) {
  if (!t->is_clicked) {
    t->is_flagged ? b->decrement_flags() : b->increment_flags();
    t->is_flagged = !t->is_flagged;
    attron(COLOR_PAIR(t->is_flagged ? COLOR_WHITE : COLOR_CYAN));
    mvaddch(io_y(t->get_y()), io_x(t->get_x()), t->is_flagged ? '<' : '.');
    attroff(COLOR_PAIR(t->is_flagged ? COLOR_WHITE : COLOR_CYAN));
  }
}

bool on_middle_click(game_board *b, tile *t) {
  bool defeated = false;
  std::vector<tile *> v;

  if (t->is_clicked && t->get_value() == b->check_tile_neighbors_flags(t->get_x(), t->get_y())) {
    v = b->get_tile_neighbors(t->get_x(), t->get_y());
    
    for (tile *tl : v) {
      if (!tl->is_flagged) {
        defeated = on_tile_click(b, tl->get_x(), tl->get_y());
        tl->is_clicked = true;
        if (defeated) {
          return true;
        }
      }
    }
  }
  return false;
}

void highlight_unclicked_neighbors(game_board *b, tile *t) {
  std::vector<tile *> v;

  /* if (t->is_clicked && t->get_value() == b->check_tile_neighbors_flags(t->get_x(), t->get_y())) { */
    v = b->get_tile_neighbors(t->get_x(), t->get_y());
    
    for (tile *tl : v) {
      if (!tl->is_flagged && !tl->is_clicked) {
        tl->is_highlighted = true;
        attron(COLOR_PAIR(COLOR_YELLOW));
        mvaddch(io_y(tl->get_y()), io_x(tl->get_x()), '.');
        attroff(COLOR_PAIR(COLOR_YELLOW));
      }
    }
  //}
}

void unhighlight(game_board *b) {
  int x, y;

  for (y = 0; y < BOARD_Y; y++) {
    for (x = 0; x < BOARD_X; x++) {
      if (b->board[y][x]->is_highlighted && !b->board[y][x]->is_clicked) {
        b->board[y][x]->is_highlighted = false;
        attron(COLOR_PAIR(COLOR_CYAN));
        mvaddch(io_y(b->board[y][x]->get_y()), io_x(b->board[y][x]->get_x()), '.');
        attroff(COLOR_PAIR(COLOR_CYAN));
      }
    }
  }
}

int level_select() {
  int in;

  clear();
  mvprintw(0, 0, "0");
  mvprintw(23, 0, "0");
  mvprintw(0, 79, "0");
  mvprintw(23, 79, "0");

  mvprintw(2, 27, "Please select a difficulty:");
  mvprintw(6, 36, "1: easy");
  mvprintw(8, 36, "2: medium");
  mvprintw(10, 36, "3: hard");

  do {
    in = getch() - '0';
  } while (in != 1 && in != 2 && in != 3);

  switch (in) {
    case 1:
      return EASY;
    case 2: 
      return MEDIUM;
    case 3:
      return HARD;
    default:
      return 200;
  }
}

void game_loop() {
    int in;
    int num_bombs;
    MEVENT event;
    game_board *board;
    bool defeated = false;
    bool generated = false;

    clear();

    num_bombs = level_select();

    grid_init();

    do {
    in = getch();
    if (getmouse(&event) == OK && in_bounds(x(event.x), y(event.y))) {
      if (!generated) {
        board = new game_board(num_bombs, x(event.x), y(event.y));
        generated = true;
      }

      if (event.bstate & BUTTON1_CLICKED) {
        defeated = on_tile_click(board, x(event.x), y(event.y));
      }
      else if (event.bstate & BUTTON2_PRESSED) {
        highlight_unclicked_neighbors(board, board->board[y(event.y)][x(event.x)]);
      }
      else if ((event.bstate & BUTTON2_RELEASED) || (event.bstate & BUTTON2_CLICKED)) {
        defeated = on_middle_click(board, board->board[y(event.y)][x(event.x)]);

        if (event.bstate & BUTTON2_RELEASED) {
          //unhighlight things (doesn't need to be done if press and release were in same place,
          //but if press->drag->release, then it needs to be done)
          //maybe add vector to board of highlighted tiles? or bool is_highlighted on tiles?
          unhighlight(board);
        }
      }
      else if (event.bstate & BUTTON3_CLICKED) {
        on_tile_flag(board, board->board[y(event.y)][x(event.x)]);
      }
    }
    
    
  } while (!defeated && in != 'Q' && !board->have_won());

  if (board->have_won()) {
    mvprintw(26, 0, "yay you won (click anywhere to continue)");
  }
  else {
    mvprintw(26, 0, "you lost (click anywhere to continue)");
  }
  show_grid(board);
  
  getch();
}

void title_screen() {
    int in;
    MEVENT event;

    clear();

    mvprintw(0, 0, "0");
    mvprintw(23, 0, "0");
    mvprintw(0, 79, "0");
    mvprintw(23, 79, "0");

    mvprintw(3, 6, "  __  __  ___  _  _  ___  ___ __      __ ___  ___  ___  ___  ___ ");
    mvprintw(4, 6, " |  \\/  ||_ _|| \\| || __|/ __|\\ \\    / /| __|| __|| _ \\| __|| _ \\");
    mvprintw(5, 6, " | |\\/| | | | | .` || _| \\__ \\ \\ \\/\\/ / | _| | _| |  _/| _| |   /");
    mvprintw(6, 6, " |_|  |_||___||_|\\_||___||___/  \\_/\\_/  |___||___||_|  |___||_|_\\");

    attron(A_BOLD);
    mvprintw(15, 34, "by alix noble");
    attroff(A_BOLD);
    mvprintw(17, 27, "--click anywhere to start--");

    do {
      in = getch();
    } while (getmouse(&event) != OK && in != 'Q');
}

int main(int argc, char const *argv[])
{
    int in;

    init_terminal();

    title_screen();

    do {
      game_loop();

      clear();

      mvprintw(2, 26, "Would you like to play again?");
      mvprintw(6, 38, "1: yes");
      mvprintw(8, 38, "2: no");

      do {
        in = getch() - '0';
      } while (in != 1 && in != 2);

    } while (in == 1);
    
    reset_terminal();
    return 0;
}
